Use ForceInline annotation in JIT

gita-omr opened the pull request:
- inside alwaysWorthInlining, check if method was annotated
   with ForceInline
- inside tryToInline, check if method was annotated
   with ForceInline
- apply to VectorAPI methods only for now

gita-omr commented:
@dmitry-ten could you please take care of the JIT Server side?

gita-omr commented:
@vijaysun-omr @jdmpapin :fyi

dmitry-ten commented on runtime/compiler/env/VMJ9Server.hpp:
```diff
@@ -292,6 +292,7 @@ class TR_J9SharedCacheServerVM: public TR_J9ServerVM
 
    virtual bool shouldDelayAotLoad() override                                  { return true; }
    virtual bool isStable(int cpIndex, TR_ResolvedMethod *owningMethod, TR::Compilation *comp) override { return false; }
+   virtual bool isForceInline(TR_ResolvedMethod *method) override { return false; }
```
This method should be overridden in `TR_J9ServerVM` instead.

gita-omr replied:
Right! Fixed.

dmitry-ten reviewed (changes requested):
LGTM but the default JITServer implementation should be in a different class.
I'll implement a proper JITServer version once this PR is merged.

gita-omr commented:
Just for the record: small Vector API benchmark runs 30% faster when *all* ForceInline annotations are respected, not just for the vector methods. So we'll try to enable that soon, in a separate PR.

dmitry-ten commented:
LGTM

gita-omr commented:
Added handling in tryToInline, removed checking disableForceInlineAnnottions option, ran all the tests. I think it's ok not to check the option since the annotation is applied to vector methods only for now, but that removes dependency on OMR. I will add the check for the option when the annotations are enabled for all methods. 

Otherwise, ready for merge.

vijaysun-omr commented on runtime/compiler/env/VMJ9.h:
```diff
@@ -1328,6 +1338,8 @@ class TR_J9SharedCacheVM : public TR_J9VM
    virtual bool               needsContiguousCodeAndDataCacheAllocation()     { return true; }
    virtual bool               needRelocatableTarget()                          { return true; }
    virtual bool               isStable(int cpIndex, TR_ResolvedMethod *owningMethod, TR::Compilation *comp) { return false; }
+   virtual bool               isForceInline(TR_ResolvedMethod *method) { return false; }
```
Is it easy to enable this functionality for AOT compiles ? I assume the above code disables it. @dsouzai fyi

dsouzai replied:
Yeah it looks like it's currently disabled in AOT. The default implementation is:
```
bool
TR_J9VMBase::isForceInline(TR_ResolvedMethod *method)
   {
   return jitIsMethodTaggedWithForceInline(vmThread(),
                                           (J9Method*)method->getPersistentIdentifier());
   }
```
which is
https://github.com/eclipse-openj9/openj9/blob/9af2e37a0d993b70055f972c02e6bb3c61a0a9b3/runtime/jit_vm/cthelpers.cpp#L219-L223
which calls
https://github.com/eclipse-openj9/openj9/blob/9af2e37a0d993b70055f972c02e6bb3c61a0a9b3/runtime/util/annhelp.c#L230-L252
which looks like a ROMMethod/ROMConstantPool query. If that's the case then we don't need to do anything special for AOT because if the VM loads a ROMClass from the SCC in the second run, then they will have already ensured it is the same ROMClass as in the first run. 

@pshipton could you confirm that the code that checks the annotation above only deals with ROM* structures?

gita-omr replied:
Also a question to @tajila

tajila replied:
Yes we are only looking at the ROM classes for annotation data. That being said, if HCR occurs (which can modify annotations) we generate a new romClass, so in those cases you'd need to be aware of potential changes.

dsouzai replied:
Regarding HCR, if the class was redefined before the load then we would fail the AOT load because the class chain wouldn't match. If it happened after the load, either a runtime assumption or just the normal runtime semantics of HCR would ensure correctness.

gita-omr commented:
@dsouzai  if your assumption above is correct, does it mean that we can just remove the override of isForceInline() that returns false from the TR_J9VM class?

Then I think we can do the same for isStable() since they use the same VM code.

dsouzai commented:
> @dsouzai if your assumption above is correct, does it mean that we can just remove the override of isForceInline() that returns false from the TR_J9VM class?

If you mean the override in `TR_J9SharedCacheVM` then yeah.

gita-omr commented:
> 
> 
> > @dsouzai if your assumption above is correct, does it mean that we can just remove the override of isForceInline() that returns false from the TR_J9VM class?
> 
> If you mean the override in `TR_J9SharedCacheVM` then yeah.

Oh, yes, of course, `TR_J9SharedCacheVM`.

gita-omr commented:
If @tajila confirms I will remove the override of isForceInline from TR_J9SharedCacheVM in this PR and isStable in another.

tajila commented:
@gita-omr 
> inside alwaysWorthInlining, check if method was annotated
with ForceInline

FYI, we also added `jitIsMethodTaggedWithDontInline`

gita-omr commented:
Thanks for all the comments! 

I did not know about the “dontInline” annotation. Maybe we can implement it in another PR? I was planning to open a PR for not inlining some intrinsic candidates. I think similar to Stable and ForceInline we would like to have a command line option to disable and experiment.

tajila commented:
> I did not know about the “dontInline” annotation.

Its the same idea as the forceinline annotation. Basically, you ignore typical inlining policy and never inline it.

gita-omr commented:
Still testing latest changes...

gita-omr commented:
Tests completed.

gita-omr commented:
I will open issues:
1) Support DontInline annotation in JIT
2) Support Stable annotation with AOT in JIT

gita-omr commented:
https://github.com/eclipse-openj9/openj9/issues/13290
https://github.com/eclipse-openj9/openj9/issues/13291
https://github.com/eclipse-openj9/openj9/issues/13292

vijaysun-omr commented:
Jenkins test extended all jdk8,jdk11

vijaysun-omr commented:
Jenkins test extended aix jdk8

vijaysun-omr commented:
Review comments have been addressed and checks have passed. Merging.

keithc-ca commented:
@vijaysun-omr Why was the failure of the line-ending check ignored?

```
20:31:03  + git diff --check origin/master HEAD
20:31:03  runtime/compiler/env/VMJ9.h:929: trailing whitespace.
20:31:03  +   
20:31:03  runtime/compiler/optimizer/InlinerTempForJ9.cpp:425: trailing whitespace.
20:31:03  +   bool vectorMethod = false; 
20:31:03  runtime/compiler/optimizer/InlinerTempForJ9.cpp:439: trailing whitespace.
20:31:03  +   
20:31:03  runtime/compiler/optimizer/InlinerTempForJ9.cpp:2188: trailing whitespace.
20:31:03  +   
20:31:03  runtime/compiler/optimizer/InlinerTempForJ9.cpp:2194: trailing whitespace.
20:31:03  +   
20:31:03  runtime/compiler/optimizer/InlinerTempForJ9.cpp:2204: trailing whitespace.
20:31:03  +   
```

gita-omr commented:
Oh, probably my fault. I think this tool was failing before and I was under impression that it was something experimental. Sorry about that!

keithc-ca commented:
I'm not aware of any problems with the line-end checker. You may have written the lines with unwanted trailing whitespace, but Vijay merged them.

vijaysun-omr commented:
Apologies, I was under the same impression, that the line-end checker was experimental in some way.

keithc-ca commented:
> Apologies, I was under the same impression, that the line-end checker was experimental in some way.

Any suggestions on how to make sure no others have that impression?

gita-omr commented:
What about running it as a part of personal build? This way the issue will be detected earlier. @keithc-ca

gita-omr commented:
@AdamBrousseau ^^

keithc-ca commented:
You can enable it for any of your local clones via this command:
```
mv .git/hooks/pre-commit.sample .git/hooks/pre-commit
```
You will then have to fix that whitespace before it can be committed.

gita-omr commented:
> Any suggestions on how to make sure no others have that impression?

Was it added couple weeks ago? Maybe it was failing then and it's fixed now?

keithc-ca commented:
The extra line-end checking was merged in #13252, nearly two weeks ago. I'm not aware of any failures of the script itself.

gita-omr commented:
Oh, I see. Maybe there were some environment issues and it's all resolved now. Thanks, great tool!

vijaysun-omr commented:
@keithc-ca do you also intend to have a similar check at the OMR project (that some of the code in the OpenJ9 codebase flows in from) ?

keithc-ca commented:
I've offered to help with https://github.com/eclipse/omr/issues/2255.
